FBTest/FBReaderCommon.swift

```diff
@@ -186,82 +186,96 @@ public extension FBReader {
         return T(reader: reader, myOffset: offset)
     }
-    
-    public func makeIterator() -> AnyIterator<T> {
-        var nextIndex = 0
-        return AnyIterator<T> {
-            if(self.count <= nextIndex){
-                return nil
-            }
-            let value = self[nextIndex]
-            nextIndex += 1
-            return value
-        }
-    }
 }
 
-struct FBScalarVector<T: Scalar, R : FBReader> : Sequence {
+struct FBScalarVector<T: Scalar, R : FBReader> : Collection {
+    public let count : Int
+
     fileprivate let reader : R
     fileprivate let myOffset : Offset
     fileprivate init(reader: R, myOffset: Offset){
         self.reader = reader
         self.myOffset = myOffset
+        self.count = reader.getVectorLength(vectorOffset: myOffset)
+    }
+
+    public var startIndex: Int {
+        return 0
     }
-    public var count : Int { return reader.getVectorLength(vectorOffset: myOffset) }
     
+    public var endIndex: Int {
+        return count
+    }
+    
+    public func index(after i: Int) -> Int {
+        return i+1
+    }
+
     public subscript(i : Int) -> T? {
         return reader.getVectorScalarElement(vectorOffset: myOffset, index: i)
     }
+}
+
+public struct FBVector<T: DirectAccess, R : FBReader> : Collection {
+    public let count : Int
     
-    public func makeIterator() -> AnyIterator<T> {
-        var nextIndex = 0
-        return AnyIterator<T> {
-            if(self.count <= nextIndex){
-                return nil
-            }
-            let value = self[nextIndex]
-            nextIndex += 1
-            return value
-        }
+    fileprivate let reader : R
+    fileprivate let myOffset : Offset
+    
+    public init(count c: Int, reader: R, myOffset: Offset){
+        self.reader = reader
+        self.myOffset = myOffset
+        self.count = c
     }
-}
 
-public struct FBVector<T> : Sequence {
-    private let _generator : (Int)->T?
-    private let _count : Int
-    init(count : Int, generator : @escaping (Int)->T?) {
-        _count = count
-        _generator = generator
+    public var startIndex: Int {
+        return 0
     }
     
-    var count : Int { return _count }
-    
-    public subscript(i : Int) -> T? {
-        return _generator(i)
+    public var endIndex: Int {
+        return count
     }
     
-    public func makeIterator() -> AnyIterator<T> {
-        var nextIndex = 0
-        return AnyIterator<T> {
-            if(self.count <= nextIndex){
-                return nil
-            }
-            let value = self._generator(nextIndex)
-            nextIndex += 1
-            return value
+    public func index(after i: Int) -> Int {
+        return i+1
+    }
+    public subscript(i : Int) -> T {
+            return T(reader: reader, myOffset: reader.getVectorOffsetElement(vectorOffset: myOffset, index: i))!
+    }
+    /*
+    public subscript(i : Int) -> T? {
```

Comment:
I would rather go with this subscript.
I generally made FlatBuffersSwift much more fault tolerant specifically on read.
Rationale is if some one sends you corrupt buffer intentionally, your application should not crash.
Specifically important for BackEnd applications